### DISCUSSION TITLE
Letting Config.layout accept all the arguments allowed by Rails

### DISCUSSION THREAD

**Discussion by elia** [solidusio/solidus#3868] [New Features or Ideas]
_Cross posting from https://github.com/solidusio/solidus/issues/3866 since it's about to become the standard place for feature requests (see https://github.com/solidusio/solidus/pull/3857)._

> **Is your feature request related to a problem? Please describe.**
> 
> 
> I need to use a simplified layout for the checkout and for other places of the store. 
> It's very common to have less UI during the checkout to avoid distractions.
> 
> Currently the `Common` controller helper is the one setting the layout and the config entry gives no extension point to dynamically select the layout based on the current controller/action/request etc. 
> As we all know "decorating" a helper is a pain in the neck, and requires the module to be re-included in all the classes its included in. It wouldn't be much of a problem if we didn't have 6 controllers including that module, if we take auth-devise into account. 
> 
> **Describe the solution you'd like**
> 
> 
> I'd like to see `Spree::Config.layout` accept a proc that takes the current controller instance as an argument and/or is evaluated with `instance_eval` inside it as it's done by Rails. Even better would be to just accept anything allowed by [`ActionView::Layouts::ClassMethods#layout`](https://api.rubyonrails.org/v6.0.0/classes/ActionView/Layouts/ClassMethods.html#method-i-layout).
> 
> I would also suggest to switch the default layout name from `spree_application` to `application` as most of the store I've seen only used the solidus layout and `spree_application` is very confusing for newcomers to solidus.
> 
> 
> 
> **Describe alternatives you've considered**
> 
> 
> - conditionals inside the layout: they would work but that's yet another a workaround fighting solidus instead of making things right
> - decorating the `Common` controller helper: it's a nightmare
> - decorating all the individual controller involved in the checkout: equally overkill and not very maintainable
> 

What do you think? 😊

**Comment by kennyadsl** [comment]
I like the idea. Have you considered deprecating the layout preference and suggest to use the rails one instead?

**Reply by elia** [reply]
@kennyadsl you mean in the `prepended` block of decorators? 
If that's the case I think it would work, but would also be tough when it comes to auth-devise. 
I think in the end the idea of an extension point for setting the layout is fine… it's just limited in it's current version.

**Reply by kennyadsl** [reply]
Sorry @elia, what's the point with auth-devise?

**Comment by devchris** [comment]
@kennyadsl  @aldesantis What do you think of the current state of #3888?
